graph/pull_v2.go

```diff
@@ -0,0 +1,361 @@
+				di.layer = layerDownload
+			}
+
+			return nil
+		}
+
+		downloads[i].err = make(chan error)
+		go func(di *downloadInfo) {
+			di.err <- downloadFunc(di)
+		}(&downloads[i])
+
+	}
+
+	var layersDownloaded bool
+	for i := len(downloads) - 1; i >= 0; i-- {
+		d := &downloads[i]
+		if d.err != nil {
+			if err := <-d.err; err != nil {
+				return false, err
+			}
+		}
+		if d.layer != nil {
+			// if tmpFile is empty assume download and extracted elsewhere
+			defer os.Remove(d.tmpFile.Name())
+			defer d.tmpFile.Close()
+			d.tmpFile.Seek(0, 0)
+			if d.tmpFile != nil {
+
+				reader := progressreader.New(progressreader.Config{
+					In:        d.tmpFile,
+					Out:       out,
+					Formatter: sf,
+					Size:      int(d.size),
+					NewLines:  false,
+					ID:        stringid.TruncateID(d.img.ID),
+					Action:    "Extracting",
+				})
+
+				err = p.graph.Register(d.img, reader)
+				if err != nil {
+					return false, err
+				}
+
+				if err := p.graph.SetDigest(d.img.ID, d.digest); err != nil {
+					return false, err
+				}
+
+				// FIXME: Pool release here for parallel tag pull (ensures any downloads block until fully extracted)
+			}
+			out.Write(sf.FormatProgress(stringid.TruncateID(d.img.ID), "Pull complete", nil))
+			layersDownloaded = true
+		} else {
+			out.Write(sf.FormatProgress(stringid.TruncateID(d.img.ID), "Already exists", nil))
+		}
+	}
+
+	/*
+		TODO(tiborvass): compare with older version below:
+
+		if localDigest != remoteDigest { // this is not a verification check.
+			// NOTE(stevvooe): This is a very defensive branch and should never
+			// happen, since all manifest digest implementations use the same
+			// algorithm.
+			logrus.WithFields(
+				logrus.Fields{
+					"local":  localDigest,
+					"remote": remoteDigest,
+				}).Debugf("local digest does not match remote")
+
+			out.Write(sf.FormatStatus("", "Remote Digest: %s", remoteDigest))
+		}
+
+		out.Write(sf.FormatStatus("", "Digest: %s", localDigest))
+
+		if tag == localDigest.String() {
+			// TODO(stevvooe): Ideally, we should always set the digest so we can
+			// use the digest whether we pull by it or not. Unfortunately, the tag
+			// store treats the digest as a separate tag, meaning there may be an
+			// untagged digest image that would seem to be dangling by a user.
+
+			if err = s.SetDigest(repoInfo.LocalName, localDigest.String(), downloads[0].img.ID); err != nil {
+				return false, err
+			}
+		}
+
+		if !utils.DigestReference(tag) {
+			// only set the repository/tag -> image ID mapping when pulling by tag (i.e. not by digest)
+			if err = s.Tag(repoInfo.LocalName, tag, downloads[0].img.ID, true); err != nil {
+				return false, err
+			}
+		}
+	*/
+	if utils.DigestReference(tag) {
+		if err = p.SetDigest(localName, tag, downloads[0].img.ID); err != nil {
+			return false, err
+		}
+	} else if err = p.Tag(localName, tag, downloads[0].img.ID, true); err != nil {
+		return false, err
+	}
+
+	if verified && layersDownloaded {
```

Comment:
Most images have not been verified so we only show the messages for the official images. The new trust work will supersede this verification.